### ISSUE TITLE
ecdh_curve25519.c example - always complains required symboles not defined

### ISSUE THREAD

**Issue by maxpeng** [Mbed-TLS/mbedtls#3191]
I builded the example [ecdh_curve25519.c](https://github.com/ARMmbed/mbedtls/blob/master/programs/pkey/ecdh_curve25519.c) with the accompanied `include/mbedtls/config.h` file. The build was successfully. However when I executed the example, it always displayed following error:
```
hMBEDTLS_ECDH_C and/or MBEDTLS_ECDH_LEGACY_CONTEXT and/or
MBEDTLS_ECP_DP_CURVE25519_ENABLED and/or MBEDTLS_ENTROPY_C and/or
MBEDTLS_CTR_DRBG_C not defined
```

After checking where the required symbols are defined, I found that one of the required symbol  `MBEDTLS_ECDH_LEGACY_CONTEXT` caused that problem. `MBEDTLS_ECDH_LEGACY_CONTEXT` is defined in ` mbedtls/ecdh.h`, that file should be included early.

The code starts from line 38 should be changed from

```
#if !defined(MBEDTLS_ECDH_C) || !defined(MBEDTLS_ECDH_LEGACY_CONTEXT) || \
    !defined(MBEDTLS_ECP_DP_CURVE25519_ENABLED) || \
    !defined(MBEDTLS_ENTROPY_C) || !defined(MBEDTLS_CTR_DRBG_C)
int main( void )
{
    mbedtls_printf( "MBEDTLS_ECDH_C and/or MBEDTLS_ECDH_LEGACY_CONTEXT and/or "
                    "MBEDTLS_ECP_DP_CURVE25519_ENABLED and/or "
                    "MBEDTLS_ENTROPY_C and/or MBEDTLS_CTR_DRBG_C "
                    "not defined\n" );
    return( 0 );
}
#else

#include "mbedtls/entropy.h"
#include "mbedtls/ctr_drbg.h"
#include "mbedtls/ecdh.h"
```
To
```
#include "mbedtls/ecdh.h"

#if !defined(MBEDTLS_ECDH_C) || !defined(MBEDTLS_ECDH_LEGACY_CONTEXT) || \
    !defined(MBEDTLS_ECP_DP_CURVE25519_ENABLED) || \
    !defined(MBEDTLS_ENTROPY_C) || !defined(MBEDTLS_CTR_DRBG_C)
int main( void )
{
    mbedtls_printf( "MBEDTLS_ECDH_C and/or MBEDTLS_ECDH_LEGACY_CONTEXT and/or "
                    "MBEDTLS_ECP_DP_CURVE25519_ENABLED and/or "
                    "MBEDTLS_ENTROPY_C and/or MBEDTLS_CTR_DRBG_C "
                    "not defined\n" );
    return( 0 );
}
#else

#include "mbedtls/entropy.h"
#include "mbedtls/ctr_drbg.h"
```

**Comment by danh-arm** [comment]
Hi @maxpeng. Thanks for reporting this. There does seem to be an issue here, although the fix may be different than the one you suggest.

`ecdh.h` conditionally #undefs `MBEDTLS_ECDH_LEGACY_CONTEXT` depending on the value of `MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED`. Are you using a config.h with `MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED` defined? If not, then I don't quite see how that code fragment you highlight is being hit.

I generally don't like #undefs as it can place an ordering dependency on #includes (as is the case here), which makes the code less maintainable. Although your fix should work, I'm wondering if a better fix would be to let the user comment out the `MBEDTLS_ECDH_LEGACY_CONTEXT` define in their config when `MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED` is enabled? This can then be checked for in check_config.h and the check in ecdh.h removed. It's possible that the wider use of these flags needs looking at too.

Anyway, are you interested in submitting a patch to fix this? Then others are more likely to give their comments.

Regards

Dan.

**Comment by danh-arm** [comment]
Actually, I've just noticed this is already checked for in https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/check_config.h#L133. So I'm now not sure how the code you refer to is being hit. Is there a different symbol causing the problem?

**Comment by danh-arm** [comment]
I just checked this program in my own environment and it works OK. Could you provide more information about your environment and how to reproduce?

**Comment by maxpeng** [comment]
I download [Mbed TLS 2.16.6](https://github.com/ARMmbed/mbedtls/releases/tag/mbedtls-2.16.6), search the source file about `#define MBEDTLS_ECDH_LEGACY_CONTEXT`, it only appears in `ecdh.h` #55. That means unless we define that macro at compile time, it is only defined in `ecdh.h`. That's why I needed to include `ecdh.h` early to solve the problem.

How to replicate this issue?
1. On a ubuntu 18.04 system, type `sudo apt install libmbedtls-dev -y`.
2. Download [ecdh_curve25519.c](https://github.com/ARMmbed/mbedtls/blob/master/programs/pkey/ecdh_curve25519.c) from `master` branch.
3. Type `gcc ecdh_curve25519.c -lmbedcrypto -o ecdh_curve25519` to build `ecdh_curve25519`.
4. Execute `ecdh_curve25519`.
5. It displays the following error:
   ```
   $ ./ecdh_curve25519
   MBEDTLS_ECDH_C and/or MBEDTLS_ECDH_LEGACY_CONTEXT and/or MBEDTLS_ECP_DP_CURVE25519_ENABLED and/or MBEDTLS_ENTROPY_C and/or MBEDTLS_CTR_DRBG_C not defined
   ```

**Comment by maxpeng** [comment]
I read your comment again, it seems I misunderstood your idea in my previous comment. Defining MBEDTLS_ECDH_LEGACY_CONTEXT in config should be a good idea.

**Comment by danh-arm** [comment]
Thanks for the extra info @maxpeng. So it seems this bug is only on the 2.16 branch. The development and 2.7 branches don't have this.

The `#define MBEDTLS_ECDH_LEGACY_CONTEXT` in `ecdh.h` was introduced in commit c9c32f3f. The commit message from that says "We add the compile time flag to the module header instead of `config.h`, because this is not a standalone feature, it only enables adding new implementations in the future". So it's correct that this is not defined in config.h in the 2.16 branch, although it *is* defined in config.h on the development branch since a new implementation was subsequently added.

It appears the problem was introduced by commit 52735ef. I now think the correct solution is to move the `#include "mbedtls/ecdh.h"` in `ecdh_curve25519.c` up as you originally suggested.

Do you want to submit a fix for this?

**Comment by maxpeng** [comment]
Sure, I will create a pull request for this.

**Comment by maxpeng** [comment]
There is another discussion about this issue for branch 2.16 - [2.16 only: incorrect dependency check in the sample program ecdh_curve25519](https://github.com/ARMmbed/mbedtls/issues/3242).

**Comment by smatt123** [comment]
Hello, what is the status on this issue? Is there a recommended way to define MBEDTLS_ECDH_LEGACY_CONTEXT to fix running the example?

**Comment by gilles-peskine-arm** [comment]
In Mbed TLS 3.x, `MBEDTLS_ECDH_LEGACY_CONTEXT` has been removed. (The symbol is still around internally but it is no longer a separate configuration option.) In 2.28.x, I'm not sure, this may be a suitable workaround.

Does https://github.com/Mbed-TLS/mbedtls/pull/5725 fix this issue (for 3.x)?

**Comment by tom-cosgrove-arm** [comment]
So this is a duplicate of #4901 and #5725 should close both issues

**Comment by gilles-peskine-arm** [comment]
@tom-cosgrove-arm Well, not exactly a duplicate. We'd identified a problem in 3.0, but there's also a problem in 2.28, as shown by the fact that this issue predates 3.0.

**Comment by gilles-peskine-arm** [comment]
@tom-cosgrove-arm @tom-daubney-arm I've checked and I can confirm that we do have a problem in 2.28.

```
scripts/config.py set MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED
scripts/config.py unset MBEDTLS_ECDH_LEGACY_CONTEXT
make
programs/pkey/ecdh_curve25519
```
results in
```
MBEDTLS_ECDH_C and/or MBEDTLS_ECDH_LEGACY_CONTEXT and/or MBEDTLS_ECP_DP_CURVE25519_ENABLED and/or MBEDTLS_ENTROPY_C and/or MBEDTLS_CTR_DRBG_C not defined
```

It's a bug that the x25519 example program doesn't work with the Everest implementation of x25519.

**Comment by tom-daubney-arm** [comment]
@gilles-peskine-arm Thanks. I have started work on a backport for this.

**Comment by tom-daubney-arm** [comment]
Fixed by [#5725](https://github.com/Mbed-TLS/mbedtls/pull/5725/).
